src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers/ApiControllerAnalyzer.cs

```diff
@@ -0,0 +1,110 @@
+
+        protected static readonly DiagnosticDescriptor Rule = new(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            // noted that highlighting the syntax makes it easier to handle partial class scenarios because
+            // the analyzer will be limited to scenarios where a base type has been defined via syntax
+            context.RegisterSyntaxNodeAction(AnalyzeSyntax, CodeAnalysis.CSharp.SyntaxKind.BaseList);
+            context.RegisterSyntaxNodeAction(AnalyzeSyntax, CodeAnalysis.VisualBasic.SyntaxKind.InheritsStatement);
+        }
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var node = context.Node; // an ImportStatementSyntax for VB or a BaseListSyntax for CS
+
+            if (!node.DescendantNodes().Any())
+            {
+                return;
+            }
+
+            // the first descendent of ImportStatementSyntax will be a QualifiedNameSyntax or IdentifierNameSyntax for VB
+            // the first descendent of BaseListSyntax will be a SimpleBaseTypeSyntax for CS
+            var baseTypeNode = context.Node.DescendantNodes().First();
+
+            if (baseTypeNode.IsKind(CodeAnalysis.CSharp.SyntaxKind.SimpleBaseType) && baseTypeNode.DescendantNodes().Any())
+            {
+                // In CSharp syntax, the SimpleBaseTypeSyntax the first child and we want the QualifiedNameSyntax or IdentifierNameSyntax
+                // resolving this node must be done before we check for symbol info
+                baseTypeNode = baseTypeNode.DescendantNodes().First();
+            }
+
+            var baseTypeSymbol = context.SemanticModel.GetSymbolInfo(baseTypeNode).Symbol as INamedTypeSymbol;
+            if (baseTypeSymbol is not null
+                && !IsBaseTypeAQualifiedReferenceToApiController(baseTypeSymbol.ToDisplayString()))
+            {
+                // we found a symbol, and the symbol information tells us this isn't an ApiController
+                // bail out
+
+                // note: at this point of upgrade assistant's workflow we are expecting the System.Web.Http reference
+                // to have been removed so if we didn't find the symbol, that's okay
+                return;
+            }
+
+            if (IsBaseTypeAQualifiedReferenceToApiController(baseTypeNode.ToString())
```

Comment:
Will ToString include trivia? Probably not. But if it does, we should get the syntax w/o trivia so we don't miss the match because of a trailing comment or something like that.

Reply:
I'll add a comment that *ToString()* is intentional here because we don't want Trivia. The method *ToFullString()* would include trivia.